### PR TITLE
Add blank_index parameter for ctc_greedy_decoder

### PR DESCRIPTION
**Pull request by pvarouktsis**
This PR resolves issue #32903. 

Added an extra parameter in ctc_greedy_decoder to be able to specify the blank index. Also updated test cases and documentation.

### CODE REVIEW

**Comment by pvarouktsis** [comment]
@googlebot I signed it!

**Comment by pvarouktsis** [comment]
> Thanks for your pull request. It looks like this may be your first contribution to a Google open source project (if not, look below for help). Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
> 📝 **Please visit https://cla.developers.google.com/ to sign.**
> 
> Once you've signed (or fixed any issues), please reply here with `@googlebot I signed it!` and we'll verify it.
> 
> #### What to do if you already signed the CLA
> ##### Individual signers
> * It's possible we don't have your GitHub username or you're using a different email address on your commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> 
> ##### Corporate signers
> * Your company has a Point of Contact who decides which employees are authorized to participate. Ask your POC to be added to the group of authorized contributors. If you don't know who your Point of Contact is, direct the Google project maintainer to [go/cla#troubleshoot](http://go/cla#troubleshoot) ([Public version](https://opensource.google/docs/cla/#troubleshoot)).
> * The email used to register you as an authorized contributor must be the email used for the Git commit. Check [your existing CLA data](https://cla.developers.google.com/clas) and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * The email used to register you as an authorized contributor must also be [attached to your GitHub account](https://github.com/settings/emails).
> 
> ℹ️ **Googlers: [Go here](https://goto.google.com/prinfo/https%3A%2F%2Fgithub.com%2Ftensorflow%2Ftensorflow%2Fpull%2F39324) for more info**.

@googlebot I signed it!

**Comment by pvarouktsis** [comment]
@googlebot I signed it!

**Comment by pvarouktsis** [comment]
Hello @ebrevdo, are there any updates on this?

**Comment by gbaned** [comment]
@ebrevdo, Any update on this PR? Please. Thanks!

File: tensorflow/python/ops/ctc_ops.py [hunk 1]
```diff
@@ -308,6 +314,10 @@ def ctc_greedy_decoder(inputs, sequence_length, merge_repeated=True):
     sequence_length: 1-D `int32` vector containing sequence lengths, having size
       `[batch_size]`.
     merge_repeated: Boolean.  Default: True.
+    blank_index: (optional) Set the class index to use for the blank label.
```
Comment (ebrevdo) [inline tensorflow/python/ops/ctc_ops.py:317]: (Optional).  Default: -1.  The class index to use for the blank label.

**Review by ebrevdo — DISMISSED** [review]
approved conditional on docstring change.

File: tensorflow/python/ops/ctc_ops.py (hunk 1 above)
Comment (pvarouktsis) [reply tensorflow/python/ops/ctc_ops.py]: Thank you!

**Review by ebrevdo — DISMISSED** [review]

**Comment by ebrevdo** [comment]
Approved.  You should fix the comments that are one space past the end of line; linter requires 2 spaces.

**Review by ebrevdo — DISMISSED** [review]
thanks!

File: tensorflow/python/ops/ctc_ops.py [hunk 2]
```diff
@@ -308,6 +314,10 @@ def ctc_greedy_decoder(inputs, sequence_length, merge_repeated=True):
     sequence_length: 1-D `int32` vector containing sequence lengths, having size
       `[batch_size]`.
     merge_repeated: Boolean.  Default: True.
+    blank_index: (Optional). Default: -1. The class index to use for the blank
+      label. Negative values will start from num_classes, ie, -1 will reproduce
+      the ctc_greedy_decoder behavior of using num_classes - 1 for the blank
+      symbol, which, actually, corresponds to the default.
```
Comment (karmel) [inline tensorflow/python/ops/ctc_ops.py:320]: This docstring is confusing; can you update to more clearly state what the default is? Also, what is this used for, and when would a user want to pass this in? 

Please also add tested examples of using this parameter in the docstring: https://www.tensorflow.org/community/contribute/docs_ref

File: tensorflow/python/ops/ctc_ops.py [hunk 3]
```diff
@@ -328,6 +338,18 @@ def ctc_greedy_decoder(inputs, sequence_length, merge_repeated=True):
...
+  if blank_index is not None:
+
+    if blank_index < 0:
+      blank_index += _get_dim(inputs, 2)
+
+    if blank_index != _get_dim(inputs, 2) - 1:
+      part_before = inputs[:, :, :blank_index]
+      part_after = inputs[:, :, blank_index + 1:]
+      part_blank = inputs[:, :, blank_index:blank_index + 1]
+      inputs = array_ops.concat([part_before, part_after, part_blank], axis=2)
```
Comment (karmel) [inline tensorflow/python/ops/ctc_ops.py:351]: This is a lot of data copying/manipulation in Python, and could have a significant performance impact for users who pass in this arg. Would it be better to do this in the op itself, or in another way that doesn't have a negative effect on performance?

**Review by karmel — CHANGES_REQUESTED** [review]
for @tensorflow/api-owners

File: tensorflow/python/ops/ctc_ops.py (hunk 2 above)
Comment (pvarouktsis) [reply tensorflow/python/ops/ctc_ops.py:320]: Thank you! I 'll proceed with those changes.

File: tensorflow/python/ops/ctc_ops.py (hunk 3 above)
Comment (pvarouktsis) [reply tensorflow/python/ops/ctc_ops.py:351]: Of course! But I thought the way I implemented it, was the same (if I am right) to the ctc_loss and that wouldn't have enough impact to user performance. Despite that, I want to improve it.

**Comment by gbaned** [comment]
@pvarouktsis Any update on this PR? Please. Thanks!

**Comment by pvarouktsis** [comment]
@gbaned Sorry! I don't have any updates yet! I 'll start working as soon as possible.

**Comment by gbaned** [comment]
@pvarouktsis Any update on this PR? Please. Thanks!

File: tensorflow/python/ops/ctc_ops.py (hunk 3 above)
Comment (ebrevdo) [reply tensorflow/python/ops/ctc_ops.py:351]: To improve this; you'll want to update the CTCGreedyDecoderOp under `core/kernels/ctc_decoder_ops.cc` and `core/ops/ctc_ops.cc`

Comment (pvarouktsis) [reply tensorflow/python/ops/ctc_ops.py:351]: @ebrevdo, thank you very much!

File: tensorflow/core/ops/ctc_ops.cc [hunk 4]
```diff
@@ -103,6 +103,7 @@ REGISTER_OP("CTCGreedyDecoder")
     .Input("inputs: T")
     .Input("sequence_length: int32")
     .Attr("merge_repeated: bool = false")
+    .Attr("blank_index: int")
```
Comment (ebrevdo) [inline tensorflow/core/ops/ctc_ops.cc:106]: this is a new attr and needs a default.  the default value is -1

File: tensorflow/core/kernels/ctc_decoder_ops.cc [hunk 5]
```diff
@@ -212,9 +213,6 @@ class CTCGreedyDecoderOp : public OpKernel {
 
     log_prob_t.setZero();
 
-    // Assumption: the blank index is num_classes - 1
-    int blank_index = num_classes - 1;
```
Comment (ebrevdo) [inline tensorflow/core/kernels/ctc_decoder_ops.cc:216]: int blank_index = (blank_index_ < 0) num_classes + blank_index_ : blank_index_;

File: tensorflow/core/kernels/ctc_decoder_ops.cc [hunk 6]
```diff
@@ -226,7 +224,7 @@ class CTCGreedyDecoderOp : public OpKernel {
           int max_class_indices;
           log_prob_t(b, 0) +=
               -RowMax<T>(input_list_t[t], b, &max_class_indices);
-          if (max_class_indices != blank_index &&
```
Comment (ebrevdo) [inline tensorflow/core/kernels/ctc_decoder_ops.cc:229]: use the blank_index i put above.

**Review by ebrevdo — CHANGES_REQUESTED** [review]
New attr needs to have a default value for backwards compatibility of existing graphs.

**Comment by gbaned** [comment]
@pvarouktsis  Can you please check @ebrevdo's comments and keep us posted ? Thanks!

**Comment by pvarouktsis** [comment]
@ebrevdo thanks a lot! Did not know that I should have a default value for backwards compatibility! I' ll proceed to these changes.

**Comment by ebrevdo** [comment]
Thanks!  You will additionally need to make sure that the blank_index you
calculate is not out of bounds to avoid segfaults.

**Comment by pvarouktsis** [comment]
Alright! I appreciate your help!

**Comment by gbaned** [comment]
@pvarouktsis Any update on this PR? Please. Thanks!

**Comment by pvarouktsis** [comment]
Unfortunately not yet, but I will have an update soon.

File: tensorflow/core/kernels/ctc_decoder_ops.cc [hunk 7]
```diff
@@ -220,8 +221,10 @@ class CTCGreedyDecoderOp : public OpKernel {
 
     log_prob_t.setZero();
 
-    // Assumption: the blank index is num_classes - 1
-    int blank_index = num_classes - 1;
+    int blank_index =
+        (blank_index_ < 0) ? num_classes + blank_index_ : blank_index_;
+    OP_REQUIRES(ctx, FastBoundsCheck(blank_index, num_classes),
+                errors::InvalidArgument("blank_index is out of bounds"));
```
Comment (ebrevdo) [inline tensorflow/core/kernels/ctc_decoder_ops.cc:227]: include the value of blank_index and num_classes in the error return.

File: tensorflow/python/ops/ctc_ops.py [hunk 8]
```diff
@@ -287,12 +287,38 @@ def _CTCLossV2Grad(op, grad_loss, _):
...
+  ...                        [ -0.1, -inf, -2.3]]])
+  >>> seq_lens = tf.constant([2, 3])
+  >>> outputs = tf.nn.ctc_greedy_decoder(
+  ...     logits,
+  ...     seq_lens,
+  ...     blank_index=1)
+
+  Notes:
+
+  - Regardless of the value of merge_repeated, if an index of a
```
Comment (ebrevdo) [inline tensorflow/python/ops/ctc_ops.py:318]: use backticks to describe arguments, like `merge_repeated`

File: tensorflow/python/ops/ctc_ops.py [hunk 9]
```diff
@@ -287,12 +287,38 @@ def _CTCLossV2Grad(op, grad_loss, _):
...
+    given time and batch corresponds to the blank_index, no new
+    element is emitted.
+  - Default blank_index is `(num_classes - 1)`, unless overriden by blank_index.
```
Comment (ebrevdo) [inline tensorflow/python/ops/ctc_ops.py:321]: ditto for `blank_index`

File: tensorflow/python/ops/ctc_ops.py [hunk 10]
```diff
@@ -308,6 +334,10 @@ def ctc_greedy_decoder(inputs, sequence_length, merge_repeated=True):
     sequence_length: 1-D `int32` vector containing sequence lengths, having size
       `[batch_size]`.
     merge_repeated: Boolean.  Default: True.
+    blank_index: (Optional). Default: num_classes - 1. Define the class index to
```
Comment (ebrevdo) [inline tensorflow/python/ops/ctc_ops.py:337]: `num_classes - 1`

**Review by ebrevdo — APPROVED** [review]
a few tiny nits, overall LGTM.

File: tensorflow/python/ops/ctc_ops.py (hunk 10 above)
Comment (pvarouktsis) [reply tensorflow/python/ops/ctc_ops.py:337]: Could you please be more specific? Thanks!

Comment (ebrevdo) [reply tensorflow/python/ops/ctc_ops.py:337]: use backticks around `num_classes - 1`.

Comment (pvarouktsis) [reply tensorflow/python/ops/ctc_ops.py:337]: Thanks a lot!!

**Review by ebrevdo — APPROVED** [review]

**Comment by rthadur** [comment]
@pvarouktsis can you please check sanity build failures ?

**Comment by pvarouktsis** [comment]
Yes, will do!

**Comment by gbaned** [comment]
@pvarouktsis  Any update on this PR? Please. Thanks!

**Comment by tensorflowbutler** [comment]
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by pvarouktsis** [comment]
Yes, I'm still valid! I'll try to have an update soon!

**Comment by tensorflowbutler** [comment]
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by pvarouktsis** [comment]
Yes, I'm still valid!

**Comment by gbaned** [comment]
@pvarouktsis Any update on this PR? Please. Thanks!

**Comment by pvarouktsis** [comment]
Sorry, I do not have any updates yet!

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by pvarouktsis** [comment]
Yes, I'm still valid.

**Comment by gbaned** [comment]
@ebrevdo Can you please review this PR ? Thanks!

**Review by allenlavoie — APPROVED** [review]
Looks fine except for the `pyext.cpp_message`->`internal.python_message` API golden changes. I assume you'll need to revert those files to get the sanity tests to pass. This will also need an API review.

**Comment by pvarouktsis** [comment]
Thank you very much!

**Review by allenlavoie — APPROVED** [review]
